Cython/Build/Cache.py

```diff
@@ -0,0 +1,189 @@
+            m = hashlib.sha1(__version__.encode('UTF-8'))
+            m.update(file_hash(filename).encode('UTF-8'))
+            for x in sorted(dependencies.all_dependencies(filename)):
+                if os.path.splitext(x)[1] not in ('.c', '.cpp', '.h'):
+                    m.update(file_hash(x).encode('UTF-8'))
+            # Include the module attributes that change the compilation result
+            # in the fingerprint. We do not iterate over module.__dict__ and
+            # include almost everything here as users might extend Extension
+            # with arbitrary (random) attributes that would lead to cache
+            # misses.
+            m.update(
+                str((language, py_limited_api,np_pythran)).encode('UTF-8')
+            )
+            m.update(compilation_options.get_fingerprint().encode('UTF-8'))
+            return m.hexdigest()
+        except OSError:
+            return None
+
+    def fingerprint_file_base(self, cfile, fingerprint):
+        return join_path(self.path, "%s-%s" % (os.path.basename(cfile), fingerprint))
+
+    def gz_fingerprint_file(self, cfile, fingerprint):
+        return self.fingerprint_file_base(cfile, fingerprint) + gzip_ext
+
+
+    def zip_fingerprint_file(self, cfile, fingerprint):
+        return self.fingerprint_file_base(cfile, fingerprint) + '.zip'
+
+    def lookup_cache(self, c_file, fingerprint):
+        # Cython-generated c files are highly compressible.
+        # (E.g. a compression ratio of about 10 for Sage).
+        if not os.path.exists(self.path):
+            safe_makedirs(self.path)
+        gz_fingerprint_file = self.gz_fingerprint_file(c_file, fingerprint)
+        zip_fingerprint_file = self.zip_fingerprint_file(c_file, fingerprint)
+        if os.path.exists(gz_fingerprint_file) or os.path.exists(zip_fingerprint_file):
+            if os.path.exists(gz_fingerprint_file):
+                os.utime(gz_fingerprint_file, None)
+                with gzip_open(gz_fingerprint_file, 'rb') as g:
+                    with open(c_file, 'wb') as f:
+                        shutil.copyfileobj(g, f)
+            else:
+                os.utime(zip_fingerprint_file, None)
+                dirname = os.path.dirname(c_file)
+                with zipfile.ZipFile(zip_fingerprint_file) as z:
+                    for artifact in z.namelist():
+                        z.extract(artifact, os.path.join(dirname, artifact))
+            return True
+        return False
+
+
+    def store_to_cache(self, fingerprint, c_file, compilation_result):
+        artifacts = list(filter(None, [
+            getattr(compilation_result, attr, None)
+            for attr in ('c_file', 'h_file', 'api_file', 'i_file')]))
+        if len(artifacts) == 1:
+            fingerprint_file = self.gz_fingerprint_file(c_file, fingerprint)
+            with open(c_file, 'rb') as f:
+                with gzip_open(fingerprint_file + '.tmp', 'wb') as g:
+                    shutil.copyfileobj(f, g)
```

Comment:
This (and the renaming afterwards) seems worth extracting. I can't see a reason why copying back from the cache to the target file should behave differently. Both benefit from a temp file.